File: 0688-knight-probability-in-chessboard/0688-knight-probability-in-chessboard.js
/**
 * @param {number} n
 * @param {number} k
 * @param {number} row
 * @param {number} column
 * @return {number}
 */
var knightProbability = function(n, k, row, column) {
    const moves = [[-2, -1], [-2, 1], [-1, 2], [1, 2], [2, -1], [2, 1], [-1, -2], [1, -2]]
    const dp = {}
    return helper(row, column, k)
    
    function helper(row, column, k) {
        const key = `${row} ${column} ${k}`
        if (row < 0 || row >= n || column < 0 || column >= n) return 0
        if (k === 0) return 1
        if (key in dp) return dp[key]
        let res = 0
        for (const [rr, cc] of moves) {
            const rrr = rr + row
            const ccc = cc + column
            res += helper(rrr, ccc, k-1) / 8
        }
        return dp[key] = res
    }
};

File: 0688-knight-probability-in-chessboard/0688-knight-probability-in-chessboard.py
class Solution:
    def knightProbability(self, n: int, k: int, row: int, column: int) -> float:
        
        moves = [[-2, -1], [-2, 1], [-1, 2], [1, 2], [2, -1], [2, 1], [-1, -2], [1, -2]]
        dp = [[[0] * n for _ in range(n)] for _ in range(k+1)]
        dp[0][row][column] = 1
        
        for i in range(1, k+1):
            for r in range(n):
                for c in range(n):
                    for rr, cc in moves:
                        rrr = rr + r
                        ccc = cc + c
                        
                        if rrr < 0 or rrr >= n or ccc < 0 or ccc >= n:
                            continue
                            
                        dp[i][r][c] += dp[i-1][rrr][ccc] / 8
                        
        return sum(map(sum, dp[k]))